### DISCUSSION TITLE
Any working examples on how to use node-windows?

### DISCUSSION THREAD

**Discussion by espenrs** [coreybutler/node-windows#330] [Q&A]
Are there any working examples or best practices on how to use node-windows? The documentation is theoretical and doesn't explain specific how and where to use it. What I can understand from the documentation you create a separate script with a service object that handles installing and uninstalling and just refers to the actual script that will be the service. But what is lacking in the documentation is how do you in the running script handle events like when the service get a closing event. Should you create the exact same service object to be able to handle that?

I feel like I need to use node-windows in two places, the windows service script for actual handling the running service, and in the install script to handle the installation of the service. But the documentation doesn't explain how to use it, only what functions it have and what they do. But not how you set up these two scenarios...

**Comment by coreybutler** [comment]
The scenario you're presenting is a paradox. A running script cannot handle its own close event, because it will no longer be running once it is closed.

However; I believe the real intent of your question is to address intentional shutdowns and monitoring existing background services.

It is possible to _control_ the background service after it is created, i.e. intentionally shutdown or start the service as shown in [this example](https://github.com/coreybutler/node-windows/blob/master/example/install.js). Node-windows is not an external process monitor though. A node-windows service will monitor itself to perform automatic restarts, but the primary goal of this tool is to integrate with the windows task/service system. This allows other common tools (like Nagios, Windows Systems Center, etc) to monitor the services.

If you want to monitor the background service and respond to close events from Node, you'll need a separate process to do that. That is not something node-windows is designed to do.

**Reply by espenrs** [reply]
Not even sure if Node.js is capable of handling normal Windows events. But any program should be able to handle these events. That's why Windows sends a close event to an application when closing it instead of forcefully closing applications. Any script or application must be given a choice to close gracefully. My scripts have some cleaning to do before they close. Any native Windows service or application will get a close event from Windows when Windows is trying to close the application. If this is not the case for Node.js or node-windows that is just too bad.

The question was more like that the documentation shows examples of how to install a Node.js script as a service using node-windows, but it does not show how to handle events inside the actual running script - like a close event. Of course if Node.js and node-windows doesn't have Windows event handling then it will just have to rely on a force quit from Windows.

**Reply by coreybutler** [reply]
Node natively has the ability to intercept SIGINT, SIGTERM, etc on all operating systems (including Windows). You don't need node-windows to handle that within a script. See the [process events](https://nodejs.org/dist/latest-v18.x/docs/api/process.html#process-events).

```js
process.on('beforeExit', handlerFn)
```

I believe this is all you'd need to do. 

There may be another option. node-windows uses [child_process events](https://nodejs.org/dist/latest-v18.x/docs/api/child_process.html) to monitor your script and will send a `shutdown` event _if the main process forcibly kills it_. You should be able to listen to this in your script using:

```js
process.on('message', function (message) {
  if (message === 'shutdown') {
    // do something
  }
})
```

This will only work if `stopparentfirst` is set, so you're probably better served by the first option.
